pandas/_libs/tslibs/parsing.pyx

```diff
@@ -188,19 +207,27 @@ cdef parse_datetime_string_with_reso(date_string, freq=None, dayfirst=False,
 
 
 cpdef bint _does_string_look_like_datetime(object date_string):
-    if date_string.startswith('0'):
-        # Strings starting with 0 are more consistent with a
-        # date-like string than a number
-        return True
-
-    try:
-        if float(date_string) < 1000:
-            return False
-    except ValueError:
-        pass
+    cdef:
+        char *buf = NULL
+        char *endptr = NULL
+        Py_ssize_t length = -1
+        double converted_date
+        char first
 
-    if date_string in _not_datelike_strings:
+    if not get_string_data(date_string, &buf, &length):
         return False
+    if length >= 1:
+        first = buf[0]
+        if first == b'0':
+            # Strings starting with 0 are more consistent with a
+            # date-like string than a number
+            return True
+        elif date_string in _not_datelike_strings:
+            return False
+        else:
```

Comment:
needs comment explaining that the next three lines are a fancy/fast way to avoid a try/except.

If there is a way to avoid needing `xstrtod` dependency that would help with my pet goal of keeping `tslibs` self-contained.  Definitely don't want to re-implement the entire xstrtod, but might be worth it if xstrtod implements machinery not needed here (e.g. scientific notation)

Reply:
My original implementation was calling libc `strtod()` with some juggling to get over the fact that Python always uses `.` as decimal point while `strtod` uses locale-dependent one. This added about 50 extra lines of C code that were searching for the decimal point and replacing it with `.` (plus it might've been lacking the removal of thousands separator).

So. while _not_ using `xstrtod()` here is possible, I doubt that it fits the bill of "not increasing complexity" and "not adding new C code" 😄 

P.S. We still need scientific notation here because this function is parsing the input from a user, and who knows how users are storing floats in their data files?..

Reply:
> We still need scientific notation here because this function is parsing the input from a user

Look at the function in context.  If this pre-check doesn't reject it, it will get passed to other parsing functions, none of which will parse scientific notation into a datetime.

Reply:
Even if so, what's the point of duplicating the code? There is already optimized C code for parsing floats, why re-implement?
If anything I can only suggest moving it to a separate `.c`/`.h` files under `tslibs` and re-use those in `tokenizer` (which makes some sense, as we might also need some fast integer parsing functions, too). Thoughts?

Reply:
> There is already optimized C code for parsing floats, why re-implement?

1) If the code in question recognizes formats that we don't want to allow, then it isn't "optimized" for this use case.
2) All else equal, we'd rather code be in cython than in C.
3) To keep the dependency graph simple (Note: I'm at the high end of the caring-about-this scale)

> If anything I can only suggest moving it to a separate .c/.h files under tslibs and re-use those in tokenizer

tokenizer is a mess of its own (in fact I'm pretty sure there are not-quite-duplicate implementations of `xstrtod`) and I'm wary of scope creep in `tslibs`.

Reply:
> 2\. All else equal, we'd rather code be in cython than in C.

But that C code is _already_ there 😄 



> 3\. To keep the dependency graph simple

So what is your preference here? I don't want to duplicate code for parsing doubles (that is improper practice IMHO), so re-using `xstrtod()` is a very nice thing to do. How do we proceed with it given your feedback?